### PR TITLE
Support overrideable json Marshaler/Unmarshaler

### PR DESCRIPTION
**Pull request by abhinavdangeti**

### CODE REVIEW

**Review by Thejas-bhat — APPROVED** [review]

File: search_test.go [hunk 1]
```diff
@@ -3228,4 +3228,4 @@ func TestDateRangeTimestampQueries(t *testing.T) {
 			}
 		}
 	}
-}
\ No newline at end of file
+}
```
Comment (moshaad7) [inline search_test.go:3231]: unrelated change

File: index/upsidedown/upsidedown.go [hunk 2]
```diff
@@ -135,22 +135,22 @@ func (udc *UpsideDownCouch) loadSchema(kvreader store.KVReader) (err error) {
 }
```
Comment (moshaad7) [inline index/upsidedown/upsidedown.go:136]: This file have unrelated changes coming from go fmt, 
we have another PR raised for "go fmt" changes (https://github.com/blevesearch/bleve/pull/1874/) based off of master.
We can rebase this PR on top of that.

Comment (abhinavdangeti) [reply index/upsidedown/upsidedown.go:136]: This is just a go fmt and given that it'll be backported as is - let's leave this as is for now and rebase #1874 instead.
